### ISSUE TITLE
Homestead rejects some REGISTERs due to a THRIFT_EAGAIN error from Cassandra

### ISSUE THREAD

**Issue by rkd-msw** [Metaswitch/homestead#387]
#### Symptoms

I registered 12,500 subscribers at a rate of 80 REGISTER flows per second (which is should be within Homestead's rated capacity). Two of these failed with a 500 error from Sprout, caused by a 500 error from Homestead, where it looks like the root cause was this error:

```
Cache of subscriber's registration data failed.

Error code: Connection Error
Description: Exception: THRIFT_EAGAIN (timed out) [2]
```

This is reported at 2016-11-23 20:45:49.761 as a "Detailed event" in the attached SAS trace.

#### Impact

A small percentage of subscribers (>0.1%) will fail to register, and therefore not get service.

#### Release and environment 

CC-homestead-10.0.0-20161116-~255.00.0.ova

This is on a deployment with one Sprout and two Homesteads (as one Sprout failed to spin up), using a HSS. Each Clearwater VM has the standard specs - 2 vCPUs and 4GB RAM.

#### Steps to reproduce

Turn up a deployment. Run stress against it using the tools from https://github.com/Metaswitch/clearwater-perf-utils.

**Comment by rkd-msw** [comment]
SAS trace at \\ad\csd\bigfiles\clearwater\issue-attachments\homestead\387

**Comment by graemerobertson** [comment]
I've seen this whilst running individual calls.

This was on a system with 2 Sprout nodes which weren't clustered (https://github.com/Metaswitch/clearwater-etcd/issues/359) and 2 Homestead nodes one of which had a Cassandra that wouldn't start (https://github.com/Metaswitch/homestead/issues/390); however it's not clear that either of those problems should have caused this. SAS trace attached.

**Comment by eleanor-merry** [comment]
This should be fixed in the latest code; passing back for retest

**Comment by rkd-msw** [comment]
Fixed on 10.0.0-20161205-~255.00.0-161205.102515.

**Comment by rkd-msw** [comment]
Reopening as I've seen this again on `10.0.0-161216-~255.00.0-161216.020317`. I suspect the earlier fix just made it less frequent - previously, I couldn't get a clean stress run without hitting this error, but now I generally can.

I ran stress for 20 minutes between 22:01:17 and 22:21:24. Between 22:18:36.874 and 22:18:38.446, I saw a burst of calls failing with 404 Not Found and re-registrations failing with 500 Internal Server Error. SAS for both cases showed the log:

```
Failed to connect to Cassandra

Error text: Exception: THRIFT_EAGAIN (timed out) [2]
```

All the SAS traces I've checked that have this error show it coming from homestead-2 (never homestead-1). I also saw this log in /var/log/cassandra/system.log on homestead-2 (but not homestead-1):

`INFO  [ScheduledTasks:1] 2016-12-19 22:18:40,391 MessagingService.java:929 - READ messages were dropped in last 5000 ms: 2 for internal timeout and 0 for cross node timeout`

Some of the 404/500 errors are not because of this THRIFT_EAGAIN, but because homestead-2 takes around a second to write data to or read data from Cassandra, so Sprout times the request out. I think the previous fix here was a timeout fix, so this may be relevant, and I've attached these traces.

**Comment by pdavies** [comment]
I have encountered this again, with THRIFT_EAGAIN occurring on both homestead-1 and homestead-2 (although mostly on homestead-2 - see attached trace).
[THRIFT_EAGAIN_on_homestead-1.zip](https://github.com/Metaswitch/homestead/files/699023/THRIFT_EAGAIN_on_homestead-1.zip)

**Comment by eleanor-merry** [comment]
This should be improved in the latest build passing back for retest

**Comment by rkd-msw** [comment]
I ran 3 hours of load, making 281,080 calls of which 83 failed with 404 Not Found errors (which is a 99.97% call success rate) - and the times of those correlate with seeing lots of THRIFT_EAGAIN errors in the logs:

```
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.994 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.995 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.995 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.995 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.999 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
/var/log/homestead/homestead_20170212T010000Z.txt:12-02-2017 01:33:38.999 UTC Error cassandra_store.cpp:433: Cassandra request failed: rc=3, Exception: THRIFT_EAGAIN (timed out) [2]
```

Re-registers had about the same success rate (99.96%).

This is on 10.0.0-170211-~255.00.0-170211.020437.

I double-checked Cassandra to check that your changes have made it in, and they have:

```
cqlsh> describe homestead_cache.impu;

CREATE TABLE homestead_cache.impu (
    public_id text PRIMARY KEY,
    ims_subscription_xml text,
    is_registered boolean,
    primary_ccf text,
    primary_ecf text,
    secondary_ccf text,
    secondary_ecf text
) WITH COMPACT STORAGE
    AND bloom_filter_fp_chance = 0.01
    AND caching = '{"keys":"ALL", "rows_per_partition":"NONE"}'
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 1.0
    AND speculative_retry = '99.0PERCENTILE';
```

It looks like https://github.com/Metaswitch/homestead/commit/4689bb8379ff0db96d68a594a88632e79f8a281a has gone back to the default speculative retry behaviour (retry when over 99th percentile latency) rather than disabling speculative retry entirely - was that intentional?

**Comment by rkd-msw** [comment]
THRIFT_EAGAIN errors came in bursts, by the way - starting at 01:19:33, 01:33:38, 01:36:29, 02:40:44 and  02:44:35 and lasting for a second to a few seconds.

**Comment by rkd-msw** [comment]
This triggers the following syslog, but not an alarm:

```
Feb 13 06:54:45 stress-homestead-2 homestead[6411]: 1008 - Description: Some connections between Homestead and Cassandra applications have failed. @@Cause: This process was unable to contact at least one instance of the application it's trying to connect to, but did make some successful contact @@Effect: This process was unable to contact at least one instance of the application it's trying to connect to @@Action: (1). Check that the application this process is trying to connect to is running.(2). Check the configuration in /etc/clearwater is correct.(3). Check that this process has connectivity to the application it's trying to connect to.
```

**Comment by eleanor-merry** [comment]
I've added a speculative fix, passing back for retest

**Comment by rkd-msw** [comment]
Nope, still 99.96% call success rate with that speculative fix (https://github.com/Metaswitch/clearwater-cassandra/pull/135). I've checked that ParallelRefProcEnabled is actually being passed to Cassandra.

**Comment by eleanor-merry** [comment]
Passing to @sebrexmetaswitch for triage

**Comment by sebrexmetaswitch** [comment]
leaving as major for now, but we've agreed what we're going to do here is to make sure Sprout is sending a sensible SIP error code (sometimes it seems to be sending a 404 when it gets back 503 errors from Homestead, where is should be sending a 504)

**Comment by sebrexmetaswitch** [comment]
Passing to you to track this, as the fix for #1623 is a workaround for this. If you're happy with the fix for that, then I think we can defer this one.
